### DISCUSSION TITLE
[Share] Replace Tag

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-actions-tags#113] [Action Scripts]
## Description

Replace tag.

Please replace the following lines in the action script before applying it:

<details>
<summary>Show lines to edit</summary>

```js
// Modify this line 👇
const tagToReplace = "TAG_TO_REPLACE";

// Modify this line 👇
const tagToAdd = "TAG_TO_ADD";
```

</details>

## Action Settings
Operation: Script
Data:
```js
// Replace tags
// @author windingwind
// @link https://github.com/windingwind/zotero-tag/discussions/113
// @usage Shortcut

if (!item) {
	return "[Replace Tags] item is empty";
}

// Modify this line 👇
const tagToReplace = "TAG_TO_REPLACE";

// Modify this line 👇
const tagToAdd = "TAG_TO_ADD";

if (item.getTags().some(obj => obj.tag === tagToReplace)) {
	item.addTag(tagToAdd);
	item.removeTag(tagToReplace);
	return `[Replace Tags] replace tag ${tagToReplace} to ${tagToAdd} on ${item.getField("title")}`;
}

return `[Replace Tags] tag ${tagToReplace} does not exist on ${item.getField("title")}`;
```

**Comment by Whiseguy** [comment]
I would like to have a script which does the following for a collection
1. copy the tags from an item (for instance a book) to it children (for instances notes or links)?
2. change a substring in the URI_link (substring_a to substring_b). (the URIs which are created by the Add Attachments > Create Link to URI)
3. iterate over the collection and merge items based on title

Could this be done?

**Comment by hahaha9968** [comment]
Great！！！

**Comment by Gitnapp** [comment]
Thanks for sharing! 

Here I did some modification to better adopt my workflow by adding matching rules with desired tag group and item with empty tag. 

For anyone who may interested about it:

```javascript
// Replace tags with a matching pattern from a group, even for untagged items
// @author windingwind
// @enhanced by gitnapp
// @link https://github.com/windingwind/zotero-tag/discussions/113 
// @usage Shortcut

if (!item) {
	return "[Replace Tags] item is empty";
}

// Define the array of tags to match
const tagsToMatch = ["/done", "/pending", "/unread", "/reading"];

// The tag to add after matching and replacing
const tagToAdd = "/tocheck";

// Check if the item is untagged or if it contains any of the tags to match
if (item.getTags().length === 0 || item.getTags().some(obj => tagsToMatch.includes(obj.tag))) {
    // If the item is untagged or contains a matching tag, remove the matching tags (if any)
    item.getTags().forEach(tagObj => {
        if (tagsToMatch.includes(tagObj.tag)) {
            item.removeTag(tagObj.tag);
        }
    });

    // Add the new tag
    item.addTag(tagToAdd);

    // Return a message indicating the replacement
    // If the item was untagged, mention that in the message
    const untaggedMessage = item.getTags().length === 0 ? " for an untagged item" : "";
    return `[Replace Tags] replaced matching tags [${tagsToMatch.join(', ')}] with ${tagToAdd} on ${item.getField("title")}${untaggedMessage}`;
}

return `[Replace Tags] no matching tag found for replacement on ${item.getField("title")}`;
```

**Comment by RyanQuey** [comment]
This one is similar as well, marks all of the selected items as `read` and removes other status tags

```
// Mark all selected items as read.
// @author windingwind
// updated by RyanQuey
// @link https://github.com/windingwind/zotero-tag/discussions/113
// Also incorporated: https://github.com/windingwind/zotero-actions-tags/discussions/113#discussioncomment-8875824
// @usage Shortcut

if (!item) {
	return "[Mark as Read] item is empty";
}

// Modify this line 👇
const tagsToReplace = ["unread", "TO-READ:High-Priority", "TO-READ", "reading"];

// Modify this line 👇
const tagToAdd = "read";

// if has any of those tags, get rid of them
item.getTags().forEach(tagObj => {
	if (tagsToReplace.includes(tagObj.tag)) {
		item.removeTag(tagObj.tag);
	}
});

// no matter what, add the `read` tag
item.addTag(tagToAdd);

return `[Mark as Read] successful for ${item.getField("title")}`;
```

**Comment by thomasotterdal** [comment]
Is it possible to trigger these scripts when adding a tag? I use emojis with assigned colours (e.g. pressing "1" on the keyboard adds tag "🟧" - see image), and it would be great if one tag could replace the next (e.g. pressing "2" would add "🟩" and remove "🟧"). 
<img width="198" alt="image" src="https://github.com/user-attachments/assets/022b87f9-8067-44d6-8e51-548fb3ad40a0" />

**Comment by geogrrl** [comment]
I also have a similar script. A number of journals add asterisks to keywords (tags) and they are annoying and ruin my workflow. This script simply deletes the asterisks. I just search for * in all fields and tags, save the search, select all in the saved search, and run the script:

### Event

None
### Operation

Script
### Data

```js
// Bulk remove a character from an array of tags
// @author windingwind
// @modified by geogrrl
// @link https://github.com/windingwind/zotero-tag/discussions/113 
// @usage Shortcut

if (!item) {
    return "[Remove Asterisks] item is empty";
}

const tags = item.getTags();
const updatedTags = tags.map(tagObj => {
    return { tag: tagObj.tag.replace(/\*/g, '') };
}).filter(tagObj => tagObj.tag !== '');

item.setTags(updatedTags);

return `[Remove Asterisks] asterisks removed from tags on ${item.getField("title")}`;
```

**Comment by jasculley** [comment]
Like @thomasotterdal I am also wondering if it is possible to have the trigger event be when a new tag is created (ideally) or added to an item. I would like to combine this with the ["convert string to lowercase"](https://github.com/windingwind/zotero-actions-tags/discussions/351#:~:text=//%20Convert%20a%20string%20to%20lower%20case%0A%20%20%20%20function%20toLowerCase(str)%20%7B%0A%20%20%20%20%20%20%20%20return%20str.toLowerCase()%3B) action from [Batch Tag Operations](https://github.com/windingwind/zotero-actions-tags/discussions/351) to automatically edit tags that contain capitalization (so all of my tags remain all lowercase).
